File: src/src_sub/content/compiled/mex/simple/source/utilities/ensemble_sparsifier/compute/EnsembleSparsifier.cpp
/*
 * This file is part of the "dev_in_place" repository located at:
 * https://github.com/osuvak/dev_in_place
 * 
 * Copyright (C) 2017  Onder Suvak
 * 
 * For licensing information check the above url.
 * Please do not remove this header.
 * */

#include "EnsembleSparsifier.hpp" 

EnsembleSparsifier::EnsembleSparsifier
  (
    size_t  N ,
    double  maxDist ,
    double *ptr_sortedEnsemble ,
    double *ptr_cdf
  ) 
  : 
  N_                  ( N ) ,
  maxDist_            ( maxDist ) ,
  ptr_sortedEnsemble_ ( ptr_sortedEnsemble ) ,
  ptr_cdf_            ( ptr_cdf )
{
  samplesIndices_.clear();
}

EnsembleSparsifier::~EnsembleSparsifier()
{
  samplesIndices_.clear();
}

void EnsembleSparsifier::computeSamples()
{
  size_t i, kLast;
  double d, lastDist;
  bool flag_found;
  
  samplesIndices_.push_back(0);
  
  while (1)
  {
    kLast = samplesIndices_.back();
    
    if ( kLast == N_ - 1 )
    {
      break;
    }
    else if ( kLast + 1 == N_ - 1)
    {
      samplesIndices_.back() = N_ - 1;
      break;
    }
    
    flag_found = false;
    
    for ( i = kLast + 1 ; i < N_ ; i++ )
    {
      d = std::sqrt 
          ( 
            std::pow( ptr_sortedEnsemble_[i] - ptr_sortedEnsemble_[kLast] , 2 ) + 
            std::pow(            ptr_cdf_[i] -            ptr_cdf_[kLast] , 2 ) 
          );
          
      lastDist = d;
          
      if ( d > maxDist_ )
      {
        flag_found = true;
        samplesIndices_.push_back(i);
        break;
      }
    } // for
    
    if( !flag_found )
    {
      samplesIndices_.push_back(N_-1);
    }
  } // while
  
  if ( lastDist < maxDist_ / 4.0 )
  {
    samplesIndices_.erase( samplesIndices_.begin() + samplesIndices_.size() - 2 );
  }
}

const std::vector<size_t> &EnsembleSparsifier::returnSamplesIndices() const
{
  return samplesIndices_;
}

File: src/src_sub/content/stuff/scripts/matlab/include/classes/+os_numerical_composite/+tools/+gpc/+preliminaries/+d_20171111_2028/EnsembleSparsifier.m
classdef EnsembleSparsifier < hgsetget
%{
/*
 * This file is part of the "dev-in-place" repository located at:
 * https://github.com/osuvak/dev-in-place
 * 
 * Copyright (C) 2017  Onder Suvak
 * 
 * For licensing information check the above url.
 * Please do not remove this header.
 * */
%}

    properties(GetAccess=public,SetAccess=public)
        
        myEnsemble
        myVal
        
        thres
    end
    
    properties(GetAccess=public,SetAccess=protected)
        
        sortedEnsemble
        cdf
        
        sortedEnsembleRaw
        
        samplesIndices
        
        N
        maxDist
        
        a
        b
    end
    
    methods(Access=public)
        
        function obj = EnsembleSparsifier()
%              if 1
%                  fprintf('Constructor called for EnsembleSparsifier.\n');
%              end
        end
        
        function computePreliminaries(obj)
            obj.N = length(obj.myEnsemble);
            
            obj.sortedEnsembleRaw = sort(obj.myEnsemble);
            obj.cdf = (1:obj.N) / obj.N;
            
            obj.a = min(obj.sortedEnsembleRaw) - obj.thres;
            obj.b = max(obj.sortedEnsembleRaw) + obj.thres - obj.a;
            
            fprintf('a : %.6e\n',obj.a);
            fprintf('b : %.6e\n',obj.b);
            
            obj.sortedEnsemble = ...
                ( obj.sortedEnsembleRaw - obj.a ) / obj.b ;
            
            choice = 2;
            
            switch lower(choice)
                case 1
                    d = ...
                        [ ...
                        obj.sortedEnsemble(2:end)-obj.sortedEnsemble(1:end-1) ; ...
                        obj.cdf(2:end)-obj.cdf(1:end-1) ...
                        ];
                    d = d.^2;
                    d = sqrt(d(1,:) + d(2,:));
            
                    obj.maxDist = sum(d) / obj.myVal;
                case 2
                    obj.maxDist = 1 / obj.myVal;
            end
            
            fprintf('Your choice of myVal yields approx. %.6e as maxDist.\n',obj.maxDist);
            
        end % function
    end
        
    methods(Access=public,Abstract)
        computeSamples(obj)
    end
end